src/cpp/shared/scripting/deepbinding.cpp

```diff
@@ -0,0 +1,43 @@
+extern "C" {
+#include "lua/src/lua.h"
+#include "lua/src/lualib.h"
+#include "lua/src/lauxlib.h"
+}
+
+#include "combat/army.h"
+#include "scripting/deepbinding.h"
+#include "town/town.h"
+
+static void make_deepbound_object(lua_State *L, int ptrAddr, std::string metatable_name) {
+  lua_newtable(L);
+  lua_pushstring(L, "ptr");
+  lua_pushinteger(L, ptrAddr);
+  lua_settable(L, -3);
+  lua_getglobal(L, metatable_name.c_str());
+  lua_setmetatable(L, -2);
+}
+
+
+void deepbound_push(lua_State *L, deepbind<army*> ptrAddr) {
+  make_deepbound_object(L, (int)ptrAddr.get(), "battleStack_mt");
```

Comment:
What is the new `deepbind` class buying you here?  Couldn't you just take an `army*` argument and `reinterpret_cast<int>` to get the integer that lua needs?

Reply:
The purpose is to signal to the compiler which version of deepbound_push to invoke, and to get a compiler error if no special version has been written. Using reinterpret_cast fails both goals.